File: Algorithm_LevelOne/Combinarions/Combinarions.cc
// 77. Combinations C++
#include <vector>

class Solution {
public:
    std::vector<std::vector<int>> output;
    
    void backtrack(int start, std::vector<int> curr, int n, int k) { 
        if (curr.size() == k) {
            output.emplace_back(curr);
            return;
        }
        for (int i = start; i < n + 1; i++) {
            curr.emplace_back(i);
            backtrack(i + 1, curr, n, k);
            curr.pop_back();
        }
    }
    std::vector<std::vector<int>> combine(int n, int k) {
        std::vector<int> curr;
        backtrack(1, curr, n, k);
        return output;
    }
};

File: Algorithm_LevelOne/Combinarions/Combinarions.py
# 77. Combinations Python


class Solution(object):
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        output = []
        def backtrack(start_elem = 1, curr=[]):
            if len(curr) == k:
                output.append(curr[:])
            for i in range(start_elem, n + 1):
                curr.append(i)
                backtrack(i + 1, curr)
                curr.pop()
        backtrack()
        return output
